src/vm/typehandle.cpp

```diff
@@ -702,39 +704,57 @@ BOOL TypeHandle::CanCastTo(TypeHandle type, TypeHandlePairList *pVisited)  const
     {
         THROWS;
         GC_TRIGGERS;
+        MODE_ANY;
         INJECT_FAULT(COMPlusThrowOM());
 
         LOADS_TYPE(CLASS_DEPENDENCIES_LOADED);
     }
     CONTRACTL_END
 
     if (*this == type)
-        return(true);
+        return true;
 
-    if (IsTypeDesc())
-        return AsTypeDesc()->CanCastTo(type, pVisited);
-                
-    if (type.IsTypeDesc())
-        return(false);
+    bool isTypeDesc = IsTypeDesc();
+    if (!isTypeDesc && type.IsTypeDesc())
+        return false;
+        
+    {
+        GCX_COOP();
 
-    return AsMethodTable()->CanCastToClassOrInterface(type.AsMethodTable(), pVisited);
+        TypeHandle::CastResult result = CastCache::TryGetFromCache(*this, type);
+        if (result != TypeHandle::MaybeCast)
+        {
+            return (BOOL)result;
+        }
+
+        if (isTypeDesc)
+            return AsTypeDesc()->CanCastTo(type, pVisited);
+
+        return AsMethodTable()->CanCastToClassOrInterface(type.AsMethodTable(), pVisited);
+    }
 }
 
 #include <optsmallperfcritical.h>
 TypeHandle::CastResult TypeHandle::CanCastToNoGC(TypeHandle type)  const
```

Comment:
What about these NoGC helpers?

Reply:
Not done yet with all the removal. 
However the entry NoGC points may need to exist in some form. Perhaps renamed. That is where the cache lookup + trivial/uncached checks would happen. 

I think I have these left:
- `ObjIsInstanceOfNoGC`  (when starting with an object) we discussed already that it is hard to remove without modifying asm helpers which we do not want to disturb just yet. Besides it is really just a cache lookup now, it does not do any "work".
- `CanCastToNoGC`           (similar to above when starting with a typehandle)
- `CanCastToNoGC`  for typedesc 
 The typedesc one is a bit more involved because it mostly has uncached checks - either too simple or not safe to cache. If arrays were not typedescs, I'd not need to touch it.
- `CanCastParamNoGC`     (for type parameters, basically an inner helper to the one above)

"NoGC" is kind of no longer the main point for these. Perhaps they should be renamed to something better. 
Renaming `ObjIsInstanceOfNoGC` would cause some amount of churn since it is used in a lot of places including asm helpers. 
I also could not come up with a good replacement for "NoGC" suffix, so I left these for now. .

Reply:
> "NoGC" is kind of no longer the main point for these. Perhaps they should be renamed to something better.

`ObjIsInstanceOfCache`?

> CanCastToNoGC for typedesc
> The typedesc one is a bit more involved because it mostly has uncached checks - either too simple or not safe to cache

What are the TypeDesc cases that are not safe to cache? I would think the TypeDesc can be on the same plat as everything else.

Reply:
Someone told me that typedescs could be manufactured on the stack sometimes (byrefs or type parameters maybe?). I did not look where that may happen, just assumed that it might happen for nonarrays. 

Casting typedescs that are not arrays is a relatively rare scenario and also relatively simple. 

I think we may be ok to not cache byrefs and method ptrs. Still thinking about type parameters. Caching certainly can be added for those if it is safe and useful.

Reply:
> Someone told me that typedescs could be manufactured on the stack sometimes

I am aware of just once place that did this in `ArrayIsInstanceOfNoGC` that you have refactored.

Reply:
If that is now safe, I will try caching TypeDesc conversions as well then. 
(and get rid of CanCastParamNoGC)

Reply:
For the suffix of the cache lookup - I will use `Cached`. 
- `CanCastToCached`
- `ObjIsInstanceOfCached`